examples_tests/02.ComputeShader/main.cpp

```diff
@@ -27,11 +38,584 @@ void main()
+	core::smart_refctd_ptr<video::IPhysicalDevice> gpu = nullptr;
+	for (size_t i = 0ull; i < gpus.size(); ++i)
+	{
+		gpu = gpus.begin()[i];
+
+		bool isGPUSuitable = false;
+
+		// Todo(achal): Abstract out
+		// Queue families --need to look for compute and present families
+		{
+			const auto& queueFamilyProperties = gpu->getQueueFamilyProperties();
+
+			for (uint32_t familyIndex = 0u; familyIndex < queueFamilyProperties.size(); ++familyIndex)
+			{
+				const auto& familyProperty = queueFamilyProperties.begin() + familyIndex;
+
+				if (familyProperty->queueFlags & video::IPhysicalDevice::E_QUEUE_FLAGS::EQF_COMPUTE_BIT)
+					computeFamilyIndex = familyIndex;
+
+				if (surface->isSupported(gpu.get(), familyIndex))
+					presentFamilyIndex = familyIndex;
+
+				if ((computeFamilyIndex != ~0u) && (presentFamilyIndex != ~0u))
+				{
+					isGPUSuitable = true;
+					break;
+				}
+			}
+		}
+
+		// Since our workload is not headless compute, a swapchain is mandatory
+		if (!gpu->isSwapchainSupported())
+			isGPUSuitable = false;
+
+		// Todo(achal): Abstract it out
+		// Check if the surface is adequate
+		{
+			uint32_t surfaceFormatCount;
+			gpu->getAvailableFormatsForSurface(surface.get(), surfaceFormatCount, nullptr);
+			std::vector<video::ISurface::SFormat> surfaceFormats(surfaceFormatCount);
+			gpu->getAvailableFormatsForSurface(surface.get(), surfaceFormatCount, surfaceFormats.data());
+
+			video::ISurface::E_PRESENT_MODE availablePresentModes =
+				gpu->getAvailablePresentModesForSurface(surface.get());
+
+			video::ISurface::SCapabilities surfaceCapabilities = {};
+			if (!gpu->getSurfaceCapabilities(surface.get(), surfaceCapabilities))
+				isGPUSuitable = false;
+
+			printf("Min swapchain image count: %d\n", surfaceCapabilities.minImageCount);
+			printf("Max swapchain image count: %d\n", surfaceCapabilities.maxImageCount);
+
+			// This flag is required, for some reason, I forgot xD
+			if ((surfaceCapabilities.supportedUsageFlags & asset::IImage::EUF_STORAGE_BIT) == 0)
+				isGPUSuitable = false;
+			
+			if ((surfaceFormats.empty()) || (availablePresentModes == video::ISurface::EPM_UNKNOWN))
+				isGPUSuitable = false;
+
+			// Todo(achal): Probably a more sophisticated way to choose these
+			minSwapchainImageCount = core::min(surfaceCapabilities.minImageCount + 1u, MAX_SWAPCHAIN_IMAGE_COUNT);
+			if ((surfaceCapabilities.maxImageCount != 0u) && (minSwapchainImageCount > surfaceCapabilities.maxImageCount))
+				minSwapchainImageCount = surfaceCapabilities.maxImageCount;
+
+			surfaceFormat = surfaceFormats[0];
+			presentMode = static_cast<video::ISurface::E_PRESENT_MODE>(availablePresentModes & (1 << 0));
+			// preTransform = static_cast<nbl::video::ISurface::E_SURFACE_TRANSFORM_FLAGS>(surfaceCapabilities.currentTransform);
+			swapchainExtent = surfaceCapabilities.currentExtent;
+		}
+
+		if (isGPUSuitable) // find the first suitable GPU
+			break;
+	}
+	assert((computeFamilyIndex != ~0u) && (presentFamilyIndex != ~0u));
```

Comment:
its all in CommonAPI already, just adjust what you need to adjust

Reply:
Resolved